File: Listas, Colas y Pilas/Colas/ColasListas.java
import java.util.Scanner;

class Nodo {
    int data;
    Nodo next;

    public Nodo(int data) {
        this.data = data;
        this.next = null;
    }
}

public class ColasListas {

    static class ColaLista {
        private Nodo front, rear;

        public ColaLista() {
            this.front = null;
            this.rear = null;
        }

        public boolean isEmpty() {
            return front == null;
        }

        public void enqueue(int data) {
            Nodo nuevoNodo = new Nodo(data);

            if (rear == null) {
                front = nuevoNodo;
                rear = nuevoNodo;
                return;
            }

            rear.next = nuevoNodo;
            rear = nuevoNodo;
        }

        public int dequeue() {
            if (isEmpty()) {
                System.out.println("Error: Cola subdesbordada (Underflow)");
                return -1;
            }

            int data = front.data;
            front = front.next;

            if (front == null) {
                rear = null;
            }

            return data;
        }

        public int peek() {
            if (isEmpty()) {
                System.out.println("Cola vacía");
                return -1;
            }
            return front.data;
        }
    }

    public static void main(String[] args) {
        ColaLista cola = new ColaLista();

        cola.enqueue(10);
        cola.enqueue(20);
        cola.enqueue(30);

        System.out.println("Elemento frontal: " + cola.peek());
        System.out.println("Elimina elemento: " + cola.dequeue());
        System.out.println("Nuevo elemento frontal: " + cola.peek());
    }
}

File: Listas, Colas y Pilas/Colas/ColasListas.js
class Nodo {
    constructor(data) {
        this.data = data;
        this.next = null;
    }
}

class ColaLista {
    constructor() {
        this.front = null;
        this.rear = null;
    }

    isEmpty() {
        return this.front === null;
    }

    enqueue(data) {
        const nuevo = new Nodo(data);
        if (this.rear === null) {
            this.front = nuevo;
            this.rear = nuevo;
            return;
        }
        this.rear.next = nuevo;
        this.rear = nuevo;
    }

    dequeue() {
        if (this.isEmpty()) {
            console.log("Cola vacía (Underflow)");
            return -1;
        }
        const d = this.front.data;
        this.front = this.front.next;
        if (this.front === null) this.rear = null;
        return d;
    }

    peek() {
        if (this.isEmpty()) {
            console.log("Cola vacía");
            return -1;
        }
        return this.front.data;
    }
}

// Prueba
const cola = new ColaLista();
cola.enqueue(10);
cola.enqueue(20);
cola.enqueue(30);

console.log("Elemento frontal:", cola.peek());
console.log("Elimina:", cola.dequeue());
console.log("Nuevo frontal:", cola.peek());
